File: frontend/src/api/gacha.ts
import { USE_MOCK, api } from './client';

export type ItemRarity = 'C' | 'R' | 'E' | 'L' | 'M';

export interface GachaDefinition {
  code: string;
  name: string;
  gacha_type: 'GOLD' | 'RP' | 'SEASON' | 'LEGEND';
  cost_currency: 'GOLD' | 'RP';
  cost_single: number;
  cost_10pull: number;
  pity_hard_ceiling: number;
  rate_common?: number;
  rate_rare: number;
  rate_epic: number;
  rate_legendary: number;
  rate_mythic?: number;
  description?: string;
  limited_label?: string;
  expires_at?: string;
}

export interface GachaPity {
  gacha_code: string;
  pull_count: number;
  pity_hard_ceiling: number;
}

export interface PulledItem {
  item_code: string;
  item_name: string;
  rarity: ItemRarity;
  slot: string;
  is_new: boolean;
}

export interface GachaPullResult {
  items: PulledItem[];
  new_pity_count: number;
  ceiling_reset: boolean;
  cost_paid: number;
  currency: 'GOLD' | 'RP';
}

const MOCK_GACHA_LIST: GachaDefinition[] = [
  {
    code: 'BASIC_PULL',
    name: 'Garage 일반 뽑기',
    gacha_type: 'GOLD',
    cost_currency: 'GOLD',
    cost_single: 200,
    cost_10pull: 1800,
    pity_hard_ceiling: 0,
    rate_common: 70,
    rate_rare: 28,
    rate_epic: 2,
    rate_legendary: 0,
    rate_mythic: 0,
    description: 'C 70% · R 28% · E 2%',
  },
  {
    code: 'PREMIUM_PULL',
    name: 'Garage 프리미엄 뽑기',
    gacha_type: 'GOLD',
    cost_currency: 'GOLD',
    cost_single: 1500,
    cost_10pull: 13500,
    pity_hard_ceiling: 100,
    rate_common: 0,
    rate_rare: 65,
    rate_epic: 33,
    rate_legendary: 2,
    rate_mythic: 0,
    description: 'R 65% · E 33% · L 2%',
  },
  {
    code: 'GC_PREMIUM_PULL',
    name: '크리스탈 뽑기',
    gacha_type: 'RP',
    cost_currency: 'RP',
    cost_single: 30,
    cost_10pull: 270,
    pity_hard_ceiling: 80,
    rate_common: 0,
    rate_rare: 50,
    rate_epic: 40,
    rate_legendary: 9,
    rate_mythic: 1,
    description: 'R 50% · E 40% · L 9% · M 1%',
  },
  {
    code: 'SEASON_PULL',
    name: '시즌 한정 뽑기',
    gacha_type: 'SEASON',
    cost_currency: 'RP',
    cost_single: 25,
    cost_10pull: 225,
    pity_hard_ceiling: 60,
    rate_common: 0,
    rate_rare: 60,
    rate_epic: 30,
    rate_legendary: 9,
    rate_mythic: 1,
    description: 'R 60% · E 30% · L 9% · M 1%',
    limited_label: 'Tết 18 days',
  },
  {
    code: 'LEGEND_PULL',
    name: '전설 뽑기',
    gacha_type: 'LEGEND',
    cost_currency: 'RP',
    cost_single: 80,
    cost_10pull: 720,
    pity_hard_ceiling: 50,
    rate_common: 0,
    rate_rare: 0,
    rate_epic: 70,
    rate_legendary: 25,
    rate_mythic: 5,
    description: 'E 70% · L 25% · M 5%',
  },
];

const MOCK_PITY: Record<string, GachaPity> = {
  PREMIUM_PULL:    { gacha_code: 'PREMIUM_PULL',    pull_count: 87, pity_hard_ceiling: 100 },
  GC_PREMIUM_PULL: { gacha_code: 'GC_PREMIUM_PULL', pull_count: 42, pity_hard_ceiling: 80 },
  SEASON_PULL:     { gacha_code: 'SEASON_PULL',      pull_count: 23, pity_hard_ceiling: 60 },
  LEGEND_PULL:     { gacha_code: 'LEGEND_PULL',      pull_count: 31, pity_hard_ceiling: 50 },
};

export async function fetchGachaList(): Promise<GachaDefinition[]> {
  if (USE_MOCK) return api.delay(MOCK_GACHA_LIST, 200);
  return api.realFetch<GachaDefinition[]>('/gacha/list');
}

export async function fetchGachaPity(gachaCode: string): Promise<GachaPity | null> {
  if (USE_MOCK) return api.delay(MOCK_PITY[gachaCode] ?? null, 100);
  return api.realFetch<GachaPity>(`/gacha/pity/${gachaCode}`);
}

export interface GachaPullLogEntry {
  gacha_code: string;
  is_10_pull: boolean;
  picked_rarity: ItemRarity;
  picked_item_code: string;
  was_pity_hit: boolean;
  cost_currency: string;
  cost_amount: number | null;
  pulled_at: string;
}

export async function fetchGachaPullLog(
  limit = 50,
  offset = 0,
): Promise<GachaPullLogEntry[]> {
  if (USE_MOCK) return api.delay([], 100);
  return api.realFetch<GachaPullLogEntry[]>(`/gacha/log?limit=${limit}&offset=${offset}`);
}

export async function pullGacha(
  gachaCode: string,
  is10Pull: boolean,
): Promise<GachaPullResult> {
  if (USE_MOCK) {
    const rarities: ItemRarity[] = is10Pull
      ? ['M', 'L', 'L', 'E', 'E', 'E', 'E', 'R', 'R', 'R']
      : ['E'];
    const items: PulledItem[] = rarities.map((r, i) => ({
      item_code: `MOCK_ITEM_${i}`,
      item_name: `Mock Item ${i + 1}`,
      rarity: r,
      slot: 'HELMET',
      is_new: i === 0,
    }));
    return api.delay({ items, new_pity_count: 0, ceiling_reset: true, cost_paid: is10Pull ? 13500 : 1500, currency: 'GOLD' }, 400);
  }
  const params = new URLSearchParams({ gacha_code: gachaCode, is_10_pull: String(is10Pull) });
  return api.realFetch<GachaPullResult>(`/gacha/pull?${params}`, { method: 'POST' });
}

File: backend/app/routers/gacha.py
import uuid

import httpx
from fastapi import APIRouter, Depends, HTTPException

from ..deps import verify_user_session
from ..engine_client import engine_client

router = APIRouter(prefix="/gacha", tags=["가챠 (Gacha)"])


_CURRENCY_MAP = {"GP": "GOLD", "GC": "RP"}


def _gacha_type(g: dict) -> str:
    currency = g.get("cost_currency", "GP")
    if currency == "GP":
        return "GOLD"
    if g.get("collection_filter"):
        return "SEASON"
    if (g.get("cost_per_pull") or 0) >= 80:
        return "LEGEND"
    return "RP"


def _to_frontend(g: dict) -> dict:
    return {
        "code": g.get("gacha_code", ""),
        "name": g.get("display_name", ""),
        "gacha_type": _gacha_type(g),
        "cost_currency": _CURRENCY_MAP.get(g.get("cost_currency", "GP"), g.get("cost_currency", "GP")),
        "cost_single": g.get("cost_per_pull", 0),
        "cost_10pull": g.get("cost_per_10_pull", 0),
        "pity_hard_ceiling": g.get("pity_threshold") or 0,
        "description": g.get("description"),
        "limited_label": None,
        "expires_at": g.get("ends_at"),
    }


@router.get("/list")
async def list_gacha(
    _uid: uuid.UUID = Depends(verify_user_session),
) -> list[dict]:
    raw = await engine_client.get_gacha_list()
    return [_to_frontend(g) for g in raw]


@router.get("/log")
async def get_pull_log(
    limit: int = 50,
    offset: int = 0,
    uid: uuid.UUID = Depends(verify_user_session),
) -> list[dict]:
    return await engine_client.get_gacha_pull_log(str(uid), limit=limit, offset=offset)


@router.get("/pity/{gacha_code}")
async def get_pity(
    gacha_code: str,
    uid: uuid.UUID = Depends(verify_user_session),
) -> dict:
    raw = await engine_client.get_gacha_pity(str(uid), gacha_code)
    # Engine returns pity_count; frontend expects pull_count + pity_hard_ceiling.
    # Fetch gacha list to get the threshold for this code.
    gacha_list = await engine_client.get_gacha_list()
    threshold = next(
        (g.get("pity_threshold") or 0 for g in gacha_list if g.get("gacha_code") == gacha_code),
        0,
    )
    return {
        "gacha_code": raw.get("gacha_code", gacha_code),
        "pull_count": raw.get("pity_count", 0),
        "pity_hard_ceiling": threshold,
    }


@router.get("/eligibility/{gacha_code}")
async def check_gacha_eligibility(
    gacha_code: str,
    uid: uuid.UUID = Depends(verify_user_session),
) -> dict:
    try:
        return await engine_client.check_gacha_eligibility(str(uid), gacha_code)
    except httpx.HTTPStatusError as e:
        raise HTTPException(
            status_code=e.response.status_code, detail=e.response.json().get("detail", str(e))
        ) from None


@router.post("/pull")
async def pull_gacha(
    gacha_code: str,
    is_10_pull: bool = False,
    uid: uuid.UUID = Depends(verify_user_session),
) -> dict:
    try:
        raw = await engine_client.pull_gacha(
            user_uuid=str(uid),
            gacha_code=gacha_code,
            is_10_pull=is_10_pull,
        )
    except httpx.HTTPStatusError as e:
        raise HTTPException(
            status_code=e.response.status_code, detail=e.response.json().get("detail", str(e))
        ) from None

    # Engine returns `results` with item_code/rarity only; frontend needs `items` with names.
    items = [
        {
            "item_code": r.get("item_code", ""),
            "item_name": r.get("item_code", ""),
            "rarity": r.get("rarity", "C"),
            "slot": "",
            "is_new": False,
        }
        for r in raw.get("results", [])
    ]
    return {
        "items": items,
        "new_pity_count": raw.get("pity_count_after", 0),
        "ceiling_reset": False,
        "cost_paid": raw.get("cost_amount", 0),
        "currency": _CURRENCY_MAP.get(raw.get("cost_currency", "GP"), raw.get("cost_currency", "GP")),
    }
